### PR TITLE
Fcds probabilistic

### PR DESCRIPTION
**Pull request by acud**
TODO:
- [ ] deserialise free slot counters on leveldb metastore ctor + test

### CODE REVIEW

File: storage/fcds/test/store.go [hunk 1]
```diff
@@ -102,6 +102,92 @@ func RunAll(t *testing.T, newStoreFunc func(t *testing.T) (fcds.Storer, func()))
...
 		RunIterator(t, newStoreFunc)
 	})
+
+	t.Run("next shard", func(t *testing.T) {
+		runNextShard(t, newStoreFunc)
+	})
+}
+
+// RunNextShard runs the test scenario for NextShard selection
+func runNextShard(t *testing.T, newStoreFunc func(t *testing.T) (fcds.Storer, func())) {
```
Comment (acud) [inline storage/fcds/test/store.go:255]: @janos @zelig I am kind of not inclined to have this test so much. i have a feeling that the test mainly checks that the random number generator will generate the correct `magic` for a given seed. WDYT?

Comment (janos) [reply storage/fcds/test/store.go:255]: @acud I think that we should tests for this behaviour, this or in some other form.

File: chunk/chunk.go [hunk 2]
```diff
@@ -119,6 +120,20 @@ func (a *Address) UnmarshalJSON(value []byte) error {
 	return nil
 }
 
+func (a *Address) UnmarshalString(s string) error {
+	v, err := hex.DecodeString(s)
+	if err != nil {
+		return err
+	}
+
+	if len(v) != AddressLength {
```
Comment (janos) [inline chunk/chunk.go:129]: Will this work with encrypted addresses?

File: storage/fcds/fcds.go [hunk 3]
```diff
@@ -357,9 +363,45 @@ func (s *Store) shardHasFreeOffsets(shard uint8) (has bool) {
 	return has
 }
 
-// getShard returns a shard number for the chunk address.
-func getShard(addr chunk.Address) (shard uint8) {
-	return addr[len(addr)-1] % shardCount
+// NextShard gets the next shard to write to.
+// uses weighted probability to choose the next shard.
```
Comment (janos) [inline storage/fcds/fcds.go]: uses -> Uses

File: storage/fcds/fcds.go [hunk 4]
```diff
@@ -357,9 +363,45 @@ func (s *Store) shardHasFreeOffsets(shard uint8) (has bool) {
...
+		// no empty offsets - they all must be considered equally as having at least
+		// one empty slot
+		intervalString += fmt.Sprintf("[%d %d) ", sum, sum+v.Slots+1)
+		sum += v.Slots + 1
+	}
+
+	// do some magic
+	magic := int64(rand.Intn(int(sum)))
+	intervalString = fmt.Sprintf("magic %d, intervals ", magic) + intervalString
+	fmt.Println(intervalString)
```
Comment (janos) [inline storage/fcds/fcds.go:466]: Remove fmt.Println

File: storage/fcds/fcds.go [hunk 5]
```diff
@@ -357,9 +363,45 @@ func (s *Store) shardHasFreeOffsets(shard uint8) (has bool) {
...
+	for _, v := range slots {
+		movingSum += v.Slots + 1
+		if magic < movingSum {
+			// we've reached the shard with the correct id
+			return v.Shard
+		}
+	}
+	//TODO: this is probably wrong
```
Comment (janos) [inline storage/fcds/fcds.go]: It may be better to return error by this function in this case.

File: storage/fcds/leveldb/leveldb.go [hunk 6]
```diff
@@ -30,17 +33,24 @@ var _ fcds.MetaStore = new(MetaStore)
...
-	db *leveldb.DB
+	db   *leveldb.DB
+	free map[uint8]int64 // free slots cardinality
+	mtx  sync.RWMutex    // synchronise free slots
 }
 
 // NewMetaStore returns new MetaStore at path.
 func NewMetaStore(path string) (s *MetaStore, err error) {
 	db, err := leveldb.OpenFile(path, &opt.Options{})
+
```
Comment (janos) [inline storage/fcds/leveldb/leveldb.go]: Why a new line?

File: storage/fcds/leveldb/leveldb.go [hunk 7]
```diff
@@ -30,17 +33,24 @@ var _ fcds.MetaStore = new(MetaStore)
...
 	if err != nil {
 		return nil, err
 	}
+
+	// todo: try to get and deserialize the free map from the persisted value on disk
```
Comment (janos) [inline storage/fcds/leveldb/leveldb.go]: Should this be implemented and have tests in order to merge this PR?

**Review by janos — COMMENTED** [review]
Thanks @acud, very nice, but some tests are failing:

```
--- FAIL: TestFCDS (0.38s)
    --- FAIL: TestFCDS/next_shard (0.06s)
        store.go:187: expected next shard value to be 1 but got 3
FAIL
FAIL    github.com/ethersphere/swarm/storage/fcds/leveldb       0.744s

--- FAIL: TestFCDS (0.17s)
    --- FAIL: TestFCDS/next_shard (0.03s)
        store.go:187: expected next shard value to be 1 but got 3
FAIL
FAIL    github.com/ethersphere/swarm/storage/fcds/mem   0.374s

--- FAIL: TestFCDS (0.02s)
    --- FAIL: TestFCDS/next_shard (0.00s)
panic: 0 [recovered]
        panic: 0

goroutine 1667 [running]:
testing.tRunner.func1.1(0x437ef20, 0x447f208)
        /usr/local/go/src/testing/testing.go:941 +0x3d0
testing.tRunner.func1(0xc00012e120)
        /usr/local/go/src/testing/testing.go:944 +0x3f9
panic(0x437ef20, 0x447f208)
        /usr/local/go/src/runtime/panic.go:967 +0x15d
github.com/ethersphere/swarm/storage/fcds/test.runNextShard(0xc00012e120, 0x441ef10)
        /Users/janos/go/src/github.com/ethersphere/swarm/storage/fcds/test/store.go:177 +0xafd
github.com/ethersphere/swarm/storage/fcds/test.RunAll.func5(0xc00012e120)
        /Users/janos/go/src/github.com/ethersphere/swarm/storage/fcds/test/store.go:107 +0x34
testing.tRunner(0xc00012e120, 0xc000056000)
        /usr/local/go/src/testing/testing.go:992 +0xdc
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1043 +0x357
FAIL    github.com/ethersphere/swarm/storage/fcds/mock  0.239s
```

File: storage/fcds/fcds.go (hunk 4 above)
Comment (acud) [reply storage/fcds/fcds.go:466]: will remove before merge as the unit test is a moving target and this helps to print the intervals to double check everything is in order (and to review :) )

File: chunk/chunk.go (hunk 2 above)
Comment (acud) [reply chunk/chunk.go:129]: i still need to check on this :+1:

**Pull request closed without merging**